File: archive/ts-poc/src/cli/hook-wrap.ts
import { spawn } from 'node:child_process'
import { Buffer } from 'node:buffer'
import { openDb, defaultDbPath } from '../db/index.js'
import { appendToOutbox } from '../db/outbox.js'
import {
  buildPayloadFromHookInput,
  parseHookInput,
} from '../adapter/claude-hook-input.js'
import type { HookEventName } from '../schema/hook-event.js'

const FALLBACK_EVENT: HookEventName = 'PreToolUse'

export interface HookWrapOptions {
  hookName: string
  command: string[]            // 원본 hook (argv). 비어 있으면 no-op (모니터링만)
  fallbackEvent?: HookEventName
  dbPath?: string
  recordToolArgs?: boolean
  customTags?: Record<string, string>
}

export interface HookWrapResult {
  exitCode: number
  outboxId: number | null      // null이면 outbox 기록 실패 (그래도 hook은 통과)
}

// stdin 전체를 동기적으로 읽기 (작은 JSON 한 번)
async function readAllStdin(): Promise<string> {
  const chunks: Buffer[] = []
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer)
  }
  return Buffer.concat(chunks).toString('utf8')
}

export async function runHookWrap(opts: HookWrapOptions): Promise<HookWrapResult> {
  const startedAt = Date.now()

  // 1. stdin을 buffer로 한 번 읽음 — buddy는 hook input을 검사도 하고
  //    원본 hook에도 다시 넘겨야 하므로
  const rawInput = await readAllStdin()
  const parsed = parseHookInput(rawInput)

  // 2. 원본 hook 실행 (있을 때만). stdio passthrough로 streaming 보장.
  let exitCode = 0
  let childPid: number | undefined

  if (opts.command.length > 0) {
    const [cmd, ...args] = opts.command
    if (cmd === undefined) {
      throw new Error('hook-wrap: command[0] is undefined')
    }
    const child = spawn(cmd, args, {
      stdio: ['pipe', 'inherit', 'inherit'],
    })
    childPid = child.pid

    // child stdin에 원본 input을 다시 흘려준다
    if (child.stdin) {
      child.stdin.end(rawInput)
    }

    exitCode = await new Promise<number>((resolve) => {
      child.on('error', () => resolve(127))   // spawn 자체 실패
      child.on('exit', (code, signal) => {
        if (signal) resolve(128 + (signalToNumber(signal) ?? 0))
        else resolve(code ?? 0)
      })
    })
  }

  const finishedAt = Date.now()

  // 3. payload 빌드 + outbox 기록 (실패해도 wrapper는 hook의 exit code를 통과)
  const payload = buildPayloadFromHookInput(parsed, {
    hookName: opts.hookName,
    fallbackEvent: opts.fallbackEvent ?? FALLBACK_EVENT,
    startedAt,
    finishedAt,
    exitCode,
    pid: childPid,
    recordToolArgs: opts.recordToolArgs ?? false,
    customTags: opts.customTags,
  })

  let outboxId: number | null = null
  try {
    const db = openDb({ path: opts.dbPath ?? defaultDbPath() })
    outboxId = appendToOutbox(db, payload)
    db.close()
  } catch (err) {
    // 절대 hook을 깨지 않는다. stderr에 한 줄만.
    process.stderr.write(
      `buddy: outbox write failed (${(err as Error).message ?? 'unknown'})\n`,
    )
  }

  return { exitCode, outboxId }
}

const SIGNAL_TABLE: Record<string, number> = {
  SIGHUP: 1,
  SIGINT: 2,
  SIGQUIT: 3,
  SIGKILL: 9,
  SIGTERM: 15,
}
function signalToNumber(sig: NodeJS.Signals): number | undefined {
  return SIGNAL_TABLE[sig]
}

File: internal/hookwrap/hookwrap.go
// Package hookwrap implements the `buddy hook-wrap` command.
//
// Invariants (v0.1 spec §7.1):
//  1. stdout silence — wrapper never writes to stdout itself.
//  2. streaming passthrough — child stdout/stderr inherit parent fds.
//  3. stdin replay — buffered once, then forwarded to child.
//  4. exit code passthrough — signals encoded as 128+sig.
//  5. outbox failure does not break the hook.
//  6. empty child command = monitoring-only, exit 0.
//  7. malformed input is absorbed; wrapper itself never crashes.
package hookwrap

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"time"

	"github.com/wm-it-22-00661/buddy/internal/adapter"
	"github.com/wm-it-22-00661/buddy/internal/db"
	"github.com/wm-it-22-00661/buddy/internal/schema"
)

const fallbackEvent = schema.EventPreToolUse

// Options carries everything Run needs from the CLI layer.
type Options struct {
	HookName       string
	Command        []string
	FallbackEvent  schema.HookEventName
	DBPath         string
	RecordToolArgs bool
	CustomTags     map[string]string

	// Injectable for tests; default to real os.Stdin/Stderr.
	Stdin  io.Reader
	Stderr io.Writer
}

// Result is what the CLI uses to set its own exit code.
type Result struct {
	ExitCode int
	OutboxID int64 // 0 if outbox write failed
}

// Run executes a hook with the wrapper invariants.
func Run(ctx context.Context, opts Options) (Result, error) {
	if opts.Stdin == nil {
		opts.Stdin = os.Stdin
	}
	if opts.Stderr == nil {
		opts.Stderr = os.Stderr
	}
	if opts.FallbackEvent == "" {
		opts.FallbackEvent = fallbackEvent
	}

	startedAt := time.Now().UnixMilli()

	// 1. buffer stdin once — wrapper inspects it and replays to child.
	rawIn, err := io.ReadAll(opts.Stdin)
	if err != nil {
		// Inv 7: never crash on stdin issue. Treat as empty input.
		rawIn = nil
	}
	parsed := adapter.Parse(string(rawIn))

	// 2. spawn child (if any). stdio passthrough for streaming.
	exitCode := 0
	var childPID int
	if len(opts.Command) > 0 {
		exitCode, childPID = runChild(ctx, opts.Command, rawIn, opts.Stderr)
	}

	finishedAt := time.Now().UnixMilli()

	// 3. build payload.
	payload := adapter.Build(parsed, adapter.AdaptOptions{
		HookName:       opts.HookName,
		FallbackEvent:  opts.FallbackEvent,
		StartedAt:      startedAt,
		FinishedAt:     finishedAt,
		ExitCode:       exitCode,
		PID:            childPID,
		RecordToolArgs: opts.RecordToolArgs,
		CustomTags:     opts.CustomTags,
	})

	// 4. outbox write — failure here MUST NOT change the hook's exit code (Inv 5).
	id, writeErr := writeOutbox(opts.DBPath, &payload)
	if writeErr != nil {
		fmt.Fprintf(opts.Stderr, "buddy: outbox write failed (%v)\n", writeErr)
	}

	return Result{ExitCode: exitCode, OutboxID: id}, nil
}

func runChild(
	ctx context.Context,
	command []string,
	stdin []byte,
	wrapperStderr io.Writer,
) (exitCode, pid int) {
	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	stdinPipe, pipeErr := cmd.StdinPipe()
	if pipeErr != nil {
		fmt.Fprintf(wrapperStderr, "buddy: stdin pipe: %v\n", pipeErr)
		return 127, 0
	}

	if err := cmd.Start(); err != nil {
		fmt.Fprintf(wrapperStderr, "buddy: spawn failed: %v\n", err)
		return 127, 0
	}
	pid = cmd.Process.Pid

	// Replay stdin to child, then close.
	if len(stdin) > 0 {
		_, _ = stdinPipe.Write(stdin)
	}
	_ = stdinPipe.Close()

	if err := cmd.Wait(); err != nil {
		var ee *exec.ExitError
		if errors.As(err, &ee) {
			st := ee.ProcessState
			if st.Exited() {
				return st.ExitCode(), pid
			}
			// Signal-terminated: encode as 128 + signum (POSIX shell convention).
			if ws, ok := st.Sys().(interface{ Signal() int }); ok {
				return 128 + ws.Signal(), pid
			}
			return 128, pid
		}
		fmt.Fprintf(wrapperStderr, "buddy: child wait: %v\n", err)
		return 127, pid
	}
	return 0, pid
}

func writeOutbox(dbPath string, p *schema.HookEventPayload) (int64, error) {
	conn, err := db.Open(db.Options{Path: dbPath})
	if err != nil {
		return 0, err
	}
	defer conn.Close()

	if err := p.Validate(); err != nil {
		return 0, fmt.Errorf("validate payload: %w", err)
	}
	return db.AppendToOutbox(conn, p)
}
